test/e2e/suite/secrettemplate/secrettemplate.go

```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secrettemplate
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	"github.com/jetstack/cert-manager/pkg/feature"
+	"github.com/jetstack/cert-manager/pkg/util"
+	"github.com/jetstack/cert-manager/test/e2e/framework"
+	e2eutil "github.com/jetstack/cert-manager/test/e2e/util"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+)
+
+const (
+	issuerName = "certificate-secret-template"
+	secretName = "test-secret-template"
+)
+
+// This test ensures that the Certificates SecretTemplate is reflected on the
+// Certificate's target Secret, and is reconciled on modify events.
+var _ = framework.CertManagerDescribe("Certificate SecretTemplate", func() {
+	f := framework.NewDefaultFramework("certificates-secret-template")
+
+	// Only run tests if the SecretTemplate feature is enabled. Skip otherwise.
+	it := func(name string, testfn func()) {
+		It(name, func() {
+			fgs := os.Getenv("FEATURE_GATES")
+			if !util.Contains(strings.Split(fgs, ","), string(feature.ExperimentalSecretApplySecretTemplateControllerMinKubernetesVTODO)+"=true") {
+				framework.Skipf("skipping Certificates SecretTemplate controller test since FEATURE_GATE %s is not enabled",
+					feature.ExperimentalSecretApplySecretTemplateControllerMinKubernetesVTODO)
+				return
```

Comment:
It took me a while to realise that I had to set this environment variable for the E2E tests too.
I see that you used this pattern for the CertificateSigningRequest controller too, but I'd forgotten about that.

The GatewayAPI support, which is also feature gated, doesn't use this technique. Instead (I think) they add specific `--ginkgo.skip` arguments from the prow environment, but I think this method is better.

Even better would be if we could query cert-manager for its current feature gates. I suppose that will be possible when we implement a configuration file for the controller-manager.

I ran the  tests locally and they passed:


```
FEATURE_GATES=ExperimentalSecretApplySecretTemplateControllerMinKubernetesVTODO=true time ./devel/run-
e2e.sh --ginkgo.focus "Certificate SecretTemplate"


• [SLOW TEST:13.199 seconds]
[cert-manager] Certificate SecretTemplate
test/e2e/framework/framework.go:283
  should not remove Annotations and Labels which have been added by a third party and not present in the SecretTemplate
  test/e2e/suite/secrettemplate/secrettemplate.go:55
------------------------------
STEP: Retrieving logs for global addons
STEP: Cleaning up the provisioned globals


Ran 7 of 796 Specs in 13.943 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 789 Skipped


Ginkgo ran 1 suite in 14.750560113s
Test Suite Passed
6.74user 2.44system 0:20.05elapsed 45%CPU (0avgtext+0avgdata 46904maxresident)k
93960inputs+512outputs (2039major+75465minor)pagefaults 0swaps

```

But these new tests are rather slow, because of the use of `consistently`. 
It makes me wonder if it's necessary to test all the cases E2E. 
Would it be sufficient to test a single happy path here and rely on the unit tests for remaining coverage?

Reply:
> It makes me wonder if it's necessary to test all the cases E2E.
> Would it be sufficient to test a single happy path here and rely on the unit tests for remaining coverage?

Unfortunately, since this functionality relies on _both_ the api-server SSA _and_ cert-manager's field manager identity, it becomes impossible to unit test/integration test without loosing confidence in the correctness.